protocol/daemons/slinky/client/price_fetcher.go

```diff
@@ -0,0 +1,144 @@
+	FetchPrices(ctx context.Context) error
+}
+
+// PriceFetcherImpl implements the PriceFetcher interface.
+type PriceFetcherImpl struct {
+	marketPairFetcher      MarketPairFetcher
+	priceFeedServiceClient api.PriceFeedServiceClient
+	grpcClient             daemontypes.GrpcClient
+	pricesSocket           string
+	pricesConn             *grpc.ClientConn
+	slinky                 oracleclient.OracleClient
+	logger                 log.Logger
+}
+
+// NewPriceFetcher creates a PriceFetcher.
+func NewPriceFetcher(
+	marketPairFetcher MarketPairFetcher,
+	grpcClient daemontypes.GrpcClient,
+	pricesSocket string,
+	slinky oracleclient.OracleClient,
+	logger log.Logger) PriceFetcher {
+	return &PriceFetcherImpl{
+		marketPairFetcher: marketPairFetcher,
+		grpcClient:        grpcClient,
+		pricesSocket:      pricesSocket,
+		slinky:            slinky,
+		logger:            logger,
+	}
+}
+
+// Start initializes the underlying connections of the PriceFetcher.
+func (p *PriceFetcherImpl) Start(ctx context.Context) error {
+	cancelCtx, cf := context.WithTimeout(ctx, SlinkyPriceServerConnectionTimeout)
+	defer cf()
+	pricesConn, err := p.grpcClient.NewGrpcConnection(cancelCtx, p.pricesSocket)
+	if err != nil {
+		return err
+	}
+	p.pricesConn = pricesConn
+	p.priceFeedServiceClient = api.NewPriceFeedServiceClient(p.pricesConn)
+	return p.slinky.Start(ctx)
+}
+
+// Stop closes all open connections.
+func (p *PriceFetcherImpl) Stop() {
+	if p.pricesConn != nil {
+		_ = p.grpcClient.CloseConnection(p.pricesConn)
+	}
+	_ = p.slinky.Stop()
+}
+
+// FetchPrices pulls prices from Slinky, translates the returned data format to dydx-compatible types,
+// and sends the price updates to the index price cache via the pricefeed server.
+// It uses the MarketPairFetcher to efficiently map between Slinky's CurrencyPair primary key and dydx's
+// MarketParam (or MarketPrice) ID.
+//
+// The markets in the index price cache will only have a single index price (from slinky).
+// This is because the sidecar pre-aggregates market data.
+func (p *PriceFetcherImpl) FetchPrices(ctx context.Context) error {
+	// get prices from slinky sidecar via GRPC
+	slinkyResponse, err := p.slinky.Prices(ctx, &types.QueryPricesRequest{})
+	if err != nil {
+		return err
+	}
+
+	// update the prices keeper w/ the most recent prices for the relevant markets
+	var updates []*api.MarketPriceUpdate
+	for currencyPairString, priceString := range slinkyResponse.Prices {
+		// convert currency-pair string (index) into currency-pair object
+		currencyPair, err := oracletypes.CurrencyPairFromString(currencyPairString)
+		if err != nil {
+			return err
+		}
+		p.logger.Info("turned price pair to currency pair",
+			"string", currencyPairString,
+			"currency pair", currencyPair.String())
+
+		// get the market id for the currency pair
+		id, err := p.marketPairFetcher.GetIDForPair(currencyPair)
+		if err != nil {
+			p.logger.Info("slinky client returned currency pair not found in MarketPairFetcher",
```

Comment:
for my understanding, this happens if slinky returns a pair that doesn't exist in `x/prices`, right?

Reply:
Yes, the `Prices` endpoint on slinky returns all pairs that are configured to be fetched. Running with a bunch of extra pairs being fetched on the sidecar returns all of these, but shouldn't be considered a failure.

Potentially making this a debug log is smart since it is quite noisy currently.